bin/faucet/build.rs

```diff
@@ -1,12 +1,104 @@
+use std::path::PathBuf;
 use std::str::FromStr;
 
-fn main() -> std::io::Result<()> {
+use anyhow::{Context, Result};
+
+/// Configures environment variables for build metadata intended for extended version information.
+fn main() -> Result<()> {
     // The location of our static faucet website files.
     let static_dir = std::path::PathBuf::from_str(std::env!("CARGO_MANIFEST_DIR"))
         .unwrap()
         .join("src")
         .join("static");
     println!("cargo::rerun-if-changed={}", static_dir.to_str().expect("Valid utf-8"));
     // This makes the static files available as an embedded resource.
-    static_files::resource_dir(static_dir).build()
+    static_files::resource_dir(static_dir).build()?;
+
+    if let Some(sha) = published_git_sha() {
+        // git data is not available if in a published state, so we set them manually.
+        println!("cargo::rustc-env=VERGEN_GIT_SHA={sha}");
+        println!("cargo::rustc-env=VERGEN_GIT_BRANCH=NA (published)");
+        println!("cargo::rustc-env=VERGEN_GIT_DIRTY=");
+
+        vergen_gitcl::Emitter::new()
+    } else {
+        // In a non-published state so we can expect git instructions to work.
+        let mut emitter = vergen_gitcl::Emitter::new();
+        emitter
+            .add_instructions(&git_instructions()?)
+            .context("Adding git instructions")?;
+
+        emitter
+    }
+    .add_instructions(&cargo_instructions()?)
+    .context("Adding cargo instructions")?
+    .add_instructions(&rustc_instructions()?)
+    .context("Adding rustc instructions")?
+    .emit()
+}
+
+/// Normal git info is lost on `cargo publish`, which instead adds a file containing the SHA1 hash.
+///
+/// This function returns the short SHA value. If present, this indicates this we're in a published state.
+fn published_git_sha() -> Option<String> {
+    let cargo_vcs_info = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(".cargo_vcs_info.json");
+    if cargo_vcs_info.exists() {
+        // The file is small so reading to string is acceptable.
+        let contents = std::fs::read_to_string(cargo_vcs_info).expect("Reading vcs info");
+
+        // File format:
+        // {
+        //   "git": {
+        //     "sha1": "9d48046e9654d93a86212e77d6c92f14c95de44b"
+        //   },
+        //   "path_in_vcs": "bin/node"
+        // }
+        let offset =
+            contents.find(r#""sha1""#).expect("Searching for sha1 property") + r#""sha1""#.len();
```

Comment:
I'm wondering if I should be more defensive about these `.expect` calls. The file does exist so we should expect its format to be okay.. However maybe it would be better to just return a string like `"Failed to parse"`.

Reply:
Yeah - I'd probably be more on the defensive side here so that we can avoid panics here for w/e unforeseen reason.

Reply:
Done. Build scripts also just succeed with a warning on failure.